examples/golang/libwaku.h

```diff
@@ -0,0 +1,45 @@
+/* Generated by Nim Compiler v1.6.11 */
+#ifndef __libwaku__
+#define __libwaku__
+#define NIM_INTBITS 64
+
+#include "nimbase.h"
+#undef LANGUAGE_C
+#undef MIPSEB
+#undef MIPSEL
+#undef PPC
+#undef R3000
+#undef R4000
+#undef i386
+#undef linux
+#undef mips
+#undef near
+#undef far
+#undef powerpc
+#undef unix
+typedef struct ConfigNode ConfigNode;
+typedef struct NimStringDesc NimStringDesc;
+typedef struct TGenericSeq TGenericSeq;
+struct TGenericSeq {NI len;
+NI reserved;
+};
+struct NimStringDesc {  TGenericSeq Sup;NIM_CHAR data[SEQ_DECL_SIZE];
+};
+typedef N_CDECL_PTR(void, tyProc__5cp59bim9aJ4WupX5aVaD1Sg) (NCSTRING signal_0);
+N_LIB_PRIVATE N_NOCONV(void, signalHandler)(int sign);
+N_LIB_PRIVATE N_NIMCALL(NI, getRefcount)(void* p);
+N_LIB_IMPORT N_CDECL(NIM_BOOL, waku_new)(ConfigNode* config, NimStringDesc** jsonResp);
```

Comment:
instead of defining the API in terms of nim types, it's _generally_ better to declare it in terms of `c` types in the nim code (ie `cint`, `cstring` etc) then import these "standard" types into the other languages

Reply:
this is for the same reason as below, ie the generated Nim code is not ABI-stable this way except certain parts

Reply:
Thanks for the comments @arnetheduck !

All of these variables get translated into `C` types thanks to the `nimbase.h` header (included in `libwaku.h`.)

`nimbase.h` brings the next definitions:
  - `#  define N_LIB_IMPORT  extern`
  - `#  define N_LIB_PRIVATE __attribute__((visibility("hidden")))`
  - `#    define N_CDECL(rettype, name) rettype name`
  - `#define NIM_BOOL _Bool`, where `_Bool` is the [C99 standard](https://www.dii.uchile.cl/~daespino/files/Iso_C_1999_definition.pdf) representation of a boolean.
  - `typedef NI64 NI;` -> `typedef int64_t NI64;` -> ··· -> `typedef signed long int __int64_t;`

Personally, given that the `libwaku.h` file is auto-generated by the Nim compiler, I wouldn't touch it.

Of course, I'm open to any other option that suits better from your point of view.
Kindly let me know if the above definitions are good enough :)

Reply:
yes, but please understand: nimbase.h changes between Nim versions: if you compile this code with nim 1.6 and nim 2.0, you will not get the same result: you will get a different, incompatible, ABI: the result of this is that the consumer of the API will need to adjust as well: when they `import` the code in their code, _their_ types will change depending on which version of Nim the code was compiled with. This is highly undesireable.

The above point is why all languages tend to use C as the "lingua franca": developers have informally agreed that plain C types are good enough for this purpose, so the "pipeline" to consume something in one language from another is that one language translates its own native types to C types and the other language does the same journey in reverse.

`NimStringDesc` is a  good example: it's no longer part of nim 2.0: it has been replaced by a different type - `NimStringDesc`, `NIM_BOOL` and so on, as their names suggest, are *internal* implementation details of the nim compiler - they are not meant to be exposed to other languages or outside of Nim.

Reply:
keep in mind that exporting code from Nim is only half of the story: the code also needs to be imported on the other side.

When importing code on the other side, this is done by translating C types seen in libwaku.h to "native" types: if, between nim versions, `NIM_BOOL` changes from `int` to `char` (this is a valid change for Nim to make), the `import` into the target language also must change - there's no reason this should be necessary: the aim of a "C bindings" module is to remove this variability by explicitly using "C FFI" types that exist for this purpose on the Nim side.